### PR TITLE
Mix of code improvements

### PR DESCRIPTION
**Pull request by tjchambers**
This is a handful of code improvements re: using of ==, and removal/refactoring of complex code.

@dkubb ready for review

### CODE REVIEW

File: lib/blockscore/collection.rb [hunk 1]
```diff
@@ -133,8 +133,7 @@ def create(params = {})
     # @api public
     def retrieve(id)
       each do |item|
-        next unless item.id == id
-        return item
+        return item if item.id.eql?(id)
```
Comment (dkubb) [inline lib/blockscore/collection.rb]: If `id` is an integer, then you could use: `item.id.equal?(id)` since it's a stronger match.

Comment (dkubb) [inline lib/blockscore/collection.rb]: I realize you're only modifying one part of this class, but I question the design of this collection class. Also, I know you didn't design this so I'm speaking to whomever designed it:

I don't think it should inherit from an Array; you now have an entire super-wide interface that your object needs to adhere to. If you ever want to refactor the object you have to re-implement the dozens of methods in Array that it implements. A much better idea for a collection class is to have a has-a relationship to an Array or Hash under the hood. You can still make it Enumerable as long as you design an `#each` method (I can help with this), and you get to control *exactly* what other methods you want to expose in your public interface.

Also, given that this method does a lookup based on an `id`, and it does so by looping over the entire collection points to the wrong data structure being used. What I would do is change this to have-a `Hash` underneath, where the key is the id, and the value is the object. You can lookup the item quickly using `Hash#fetch`, and still have `#each`:

```ruby
include Enumerable

# ...

def each(&block)
  return to_enum unless block
  items.each_value(&block)
  self
end
```

File: lib/blockscore/collection.rb [hunk 2]
```diff
@@ -197,7 +196,7 @@ def new_member(instance, &blk)
     #
     # @api private
     def parent_id?(item)
-      parent.id && item.send(foreign_key) == parent.id
+      parent.id && item.send(foreign_key).eql?(parent.id)
```
Comment (dkubb) [inline lib/blockscore/collection.rb]: I would use `#equal?` here instead of `#eql?`

Comment (tjchambers) [inline lib/blockscore/collection.rb]: These id's are strings.

File: lib/blockscore/collection.rb (hunk 1 above)
Comment (tjchambers) [inline lib/blockscore/collection.rb]: None of the id's in the system are integers, they are strings (similar to uuid's). So eql? is the best I know - correct?

**Comment by tjchambers** [comment]
@dkubb rebased and ready for re-review

File: lib/blockscore/collection.rb (hunk 1 above)
Comment (dkubb) [inline lib/blockscore/collection.rb]: Yes, `#eql?` is the best you can get in this case if the `id` are string objects.

**Pull request merged**